### PR TITLE
Added deployment plan and recipe support for delpoyment plans.

### PR DESCRIPTION
**Pull request by hswami**

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Deployment/Deployment/DeploymentPlanDeploymentSource.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                             plan.Name,
+                             Steps = (from step in plan.DeploymentSteps
+                                     select new
+                                     {
+                                         Type = step.GetType().Name,
+                                         Step = step
+                                     }).ToArray()
+                         }).ToArray();
+
+            // Adding Lucene settings
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Deployment/Deployment/DeploymentPlanDeploymentSource.cs]: typo

File: src/OrchardCore.Modules/OrchardCore.Deployment/DeploymentPlanService.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        }
+
+        public async Task CreateOrUpdateDeploymentPlans(IEnumerable<DeploymentPlan> deploymentPlans)
+        {
+            foreach (var deploymentPlan in deploymentPlans)
+            {
+                _session.Save(deploymentPlan);
+            }
+
+            await _session.CommitAsync();
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Deployment/DeploymentPlanService.cs]: It shouldn't be necessary. Our scopes automatically do it.

File: src/OrchardCore.Modules/OrchardCore.Deployment/DeploymentPlanService.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                    var deploymentPlans = await deploymentPlanQuery.ListAsync();
+                    return deploymentPlans.ToDictionary(x => x.Name);
+                }).Unwrap());
+        }
+
+        public async Task<bool> DoesUserHavePermissionsAsync()
+        {
+            var user = _httpContextAccessor.HttpContext.User;
+
+            var result = await _authorizationService.AuthorizeAsync(user, Permissions.ManageDeploymentPlan);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Deployment/DeploymentPlanService.cs]: `&&` should work

File: src/OrchardCore.Modules/OrchardCore.Deployment/DeploymentPlanService.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        public DeploymentPlanService(
+            YesSql.ISession session,
+            IHttpContextAccessor httpContextAccessor,
+            IAuthorizationService authorizationService)
+        {
+            _session = session;
+            _httpContextAccessor = httpContextAccessor;
+            _authorizationService = authorizationService;
+            _deploymentPlans = new Lazy<Task<Dictionary<string, DeploymentPlan>>>(() =>
+                Task.Factory.StartNew(async () =>
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Deployment/DeploymentPlanService.cs]: no, prefer a local field instead, and check it's set of load it.

File: src/OrchardCore.Modules/OrchardCore.Deployment/Deployment/DeploymentPlanDeploymentSource.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        private readonly DeploymentPlanService _deploymentPlanService;
+
+        public DeploymentPlanDeploymentSource(DeploymentPlanService deploymentPlanService)
+        {
+            _deploymentPlanService = deploymentPlanService;
+        }
+
+        public async Task ProcessDeploymentStepAsync(DeploymentStep deploymentStep, DeploymentPlanResult result)
+        {
+            var deploymentPlanStep = deploymentStep as DeploymentPlanDeploymentStep;
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Deployment/Deployment/DeploymentPlanDeploymentSource.cs]:
```
if (deploymentStep is DeploymentPlanDeploymentStep deploymentPlanStep)
{
}```

**Comment by hswami** [comment]
Excellent feedback! Thank you.

**Pull request merged**